kafka-connect/kafka-connect/src/main/java/org/apache/iceberg/connect/data/Utilities.java

```diff
@@ -0,0 +1,247 @@
+            .impl("org.apache.hadoop.conf.Configuration")
+            .orNull()
+            .build();
+
+    if (configClass == null) {
+      LOG.info("Hadoop not found on classpath, not creating Hadoop config");
+      return null;
+    }
+
+    try {
+      Object result = DynConstructors.builder().hiddenImpl(configClass).build().newInstance();
+      BoundMethod addResourceMethod =
+          DynMethods.builder("addResource").impl(configClass, URL.class).build(result);
+      BoundMethod setMethod =
+          DynMethods.builder("set").impl(configClass, String.class, String.class).build(result);
+
+      //  load any config files in the specified config directory
+      String hadoopConfDir = config.hadoopConfDir();
+      if (hadoopConfDir != null) {
+        HADOOP_CONF_FILES.forEach(
+            confFile -> {
+              Path path = Paths.get(hadoopConfDir, confFile);
+              if (Files.exists(path)) {
+                try {
+                  addResourceMethod.invoke(path.toUri().toURL());
+                } catch (IOException e) {
+                  LOG.warn("Error adding Hadoop resource {}, resource was not added", path, e);
+                }
+              }
+            });
+      }
+
+      // set any Hadoop properties specified in the sink config
+      config.hadoopProps().forEach(setMethod::invoke);
+
+      LOG.info("Hadoop config initialized: {}", configClass.getName());
+      return result;
+    } catch (Exception e) {
+      LOG.warn(
+          "Hadoop found on classpath but could not create config, proceeding without config", e);
+    }
+    return null;
+  }
+
+  public static Object extractFromRecordValue(Object recordValue, String fieldName) {
+    List<String> fields = Splitter.on('.').splitToList(fieldName);
+    if (recordValue instanceof Struct) {
+      return valueFromStruct((Struct) recordValue, fields);
+    } else if (recordValue instanceof Map) {
+      return valueFromMap((Map<?, ?>) recordValue, fields);
+    } else {
+      throw new UnsupportedOperationException(
+          "Cannot extract value from type: " + recordValue.getClass().getName());
+    }
+  }
+
+  private static Object valueFromStruct(Struct parent, List<String> fields) {
+    Struct struct = parent;
+    for (int idx = 0; idx < fields.size() - 1; idx++) {
+      Object value = fieldValueFromStruct(struct, fields.get(idx));
+      if (value == null) {
+        return null;
+      }
+      Preconditions.checkState(value instanceof Struct, "Expected a struct type");
+      struct = (Struct) value;
+    }
+    return fieldValueFromStruct(struct, fields.get(fields.size() - 1));
+  }
+
+  private static Object fieldValueFromStruct(Struct struct, String fieldName) {
+    Field structField = struct.schema().field(fieldName);
+    if (structField == null) {
+      return null;
+    }
+    return struct.get(structField);
+  }
+
+  private static Object valueFromMap(Map<?, ?> parent, List<String> fields) {
+    Map<?, ?> map = parent;
```

Comment:
Minor: this is `Map<String, ?>`?

Reply:
Yes, the key is always the field name, so I changed this.